<div style="border: solid blue 2px; padding: 15px; margin: 10px">
  <b>Overall Summary of the Project – Final Report</b><br><br>

  Hi Michael, I’m <b>Victor Camargo</b> (<a href="https://hub.tripleten.com/u/e9cc9c11" target="_blank">TripleTen Hub profile</a>). Thanks for your detailed submission, I’ve reviewed your final project and report and I'm happy to say it is now <b>approved</b>! 🎉<br><br>

  <b>Nice work on:</b><br>
  ✔️ Merging and cleaning all four datasets with consistent column names and correct handling of missing values.<br>
  ✔️ Converting data types properly and engineering meaningful features such as <b>tenure_days</b> and <b>tenure_months</b>.<br>
  ✔️ Creating a <b>DummyClassifier</b> baseline and then testing a wide range of models (Logistic Regression, Decision Tree, Random Forest, XGBoost, LightGBM, CatBoost).<br>
  ✔️ Addressing the <b>class imbalance</b> thoughtfully using weights and <code>scale_pos_weight</code> in boosted models.<br>
  ✔️ Implementing <b>Pipelines</b> with preprocessing, which keeps the workflow clean and reproducible.<br>
  ✔️ Using both validation and cross-validation with the ROC-AUC metric, and then confirming performance on the holdout test set.<br>
  ✔️ Writing a clear conclusion, with strong reasoning for selecting <b>CatBoost</b> as the final model based on both performance and ease of handling categorical features.<br><br>

  <b>Final note:</b><br>
  You’ve built a very solid end-to-end machine learning pipeline. The structure of your work, the clarity of your results, and the quality of your reflection all show excellent understanding. <b>Congratulations on reaching this milestone — great job!</b> 🚀💪
</div>


# Final Project: Model Testing and Analysis

## Loading Libraries and Preparing Data

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.dummy import DummyClassifier

from sklearn.metrics import roc_auc_score
from sklearn.model_selection import cross_val_score

from lightgbm import LGBMClassifier
from xgboost import XGBClassifier
from catboost import CatBoostClassifier

In [2]:
contract_df = pd.read_csv('/datasets/final_provider/contract.csv')
personal_df = pd.read_csv('/datasets/final_provider/personal.csv')
internet_df = pd.read_csv('/datasets/final_provider/internet.csv')
phone_df = pd.read_csv('/datasets/final_provider/phone.csv')

In [3]:
contract_df = contract_df.rename(columns = {'customerID':'customer_id', 'BeginDate':'begin_date', 'EndDate':'end_date',
                                            'Type':'type', 'PaperlessBilling':'paperless_billing', 'PaymentMethod':'payment_method',
                                           'MonthlyCharges':'monthly_charges', 'TotalCharges':'total_charges'})
personal_df = personal_df.rename(columns = {'customerID':'customer_id', 'SeniorCitizen':'senior_citizen', 'Partner':'partner',
                                           'Dependents':'dependents'})
internet_df = internet_df.rename(columns = {'customerID':'customer_id', 'InternetService':'internet_service', 
                                            'OnlineSecurity':'online_security', 'OnlineBackup':'online_backup',
                                           'DeviceProtection':'device_protection', 'TechSupport':'tech_support',
                                           'StreamingTV':'streaming_tv', 'StreamingMovies':'streaming_movies'})
phone_df = phone_df.rename(columns = {'customerID':'customer_id', 'MultipleLines':'multiple_lines'})

In [4]:
merge1 = contract_df.merge(personal_df, on = 'customer_id', how = 'left')
merge2 = internet_df.merge(phone_df, on = 'customer_id', how = 'left')
df = merge1.merge(merge2, on = 'customer_id', how = 'left')

In [5]:
internet_columns = ['internet_service', 'online_security', 'online_backup', 'device_protection', 'tech_support', 'streaming_tv',
                   'streaming_movies']
df[internet_columns] = df[internet_columns].fillna('No internet service')

df['multiple_lines'] = df['multiple_lines'].fillna('No phone service')

In [6]:
df['total_charges'] = pd.to_numeric(df['total_charges'], errors = 'coerce')
df['total_charges'] = df['total_charges'].fillna(0)

In [7]:
df['begin_date'] = pd.to_datetime(df['begin_date'])
df['end_date'] = pd.to_datetime(df['end_date'].replace("No", pd.NaT))

In [8]:
df['partner'] = df['partner'].replace('Yes', 1)
df['partner'] = df['partner'].replace('No', 0)
df['paperless_billing'] = df['paperless_billing'].replace('Yes', 1)
df['paperless_billing'] = df['paperless_billing'].replace('No', 0)
df['dependents'] = df['dependents'].replace('Yes', 1)
df['dependents'] = df['dependents'].replace('No', 0)

In [9]:
cutoff_date = pd.to_datetime("2020-02-01")
df['tenure_days'] = (df['end_date'].fillna(cutoff_date) - df['begin_date']).dt.days
df['tenure_months'] = df['tenure_days'] // 30
df['churn'] = df['end_date'].notna().astype(int)

In [10]:
customer_ids = df['customer_id']
features = df.drop(['customer_id', 'begin_date', 'end_date', 'churn'], axis = 1)
target = df['churn']
print(features.shape, target.shape)

(7043, 19) (7043,)


In [11]:
cat_cols = features.select_dtypes(include = ['object']).columns
num_cols = ['monthly_charges', 'total_charges', 'tenure_months', 'tenure_days']

In [12]:
features_train, features_temp, target_train, target_temp = train_test_split(features, target, test_size = 0.4, random_state = 2356)
features_test, features_valid, target_test, target_valid = train_test_split(features_temp, target_temp, test_size = 0.5, random_state = 2356)
print(features_train.shape, target_train.shape)
print()
print(features_valid.shape, target_valid.shape, features_test.shape, target_test.shape)

(4225, 19) (4225,)

(1409, 19) (1409,) (1409, 19) (1409,)


Splitting our data into a training, test, and validation set at a typical 3:1:1 ratio

In [13]:
preprocessor = ColumnTransformer(
    transformers = [('num', StandardScaler(), num_cols), ('cat', OneHotEncoder(drop = 'first'), cat_cols)
                   ]
)

In [14]:
#determining class weights and counts for scale_pos_weight use in XGB and LGBM
class_counts = df['end_date'].notna().value_counts()
class_counts

False    5174
True     1869
Name: end_date, dtype: int64

In [15]:
scale_pos_weight = class_counts[0] / class_counts[1]

In [16]:
cat_boost_weights = [(len(df) / (2 * class_counts[0])), (len(df) / (2 * class_counts[1]))]

## Analysis and Evaluation of Training and Validation Data

In [17]:
models = {'Logistic Regression': LogisticRegression(class_weight = 'balanced',
                                                                    solver = 'liblinear',
                                                                    random_state = 2356
                                                   ),
          'Decision Tree': DecisionTreeClassifier(class_weight = 'balanced',
                                                  random_state = 2356
                                                 ),
          'Random Forest': RandomForestClassifier(class_weight = 'balanced',
                                                  random_state = 2356
                                                 ),
          'XGBoost': XGBClassifier(scale_pos_weight = scale_pos_weight,
                                   use_label_encoder = False,
                                   eval_metric = 'auc',
                                   random_state = 2356
                                  ),
          'LightGBM': LGBMClassifier(scale_pos_weight = scale_pos_weight,
                                     random_state = 2356
                                    ),
          'CatBoost': CatBoostClassifier(class_weights = cat_boost_weights,
                                         random_state = 2356,
                                         verbose = 0
                                        )
         }
results = {}

for name, model in models.items():
    pipeline = Pipeline([
        ('preprocessor', preprocessor), ('classifier', model)
    ])
    pipeline.fit(features_train, target_train)
    pred_prob = pipeline.predict_proba(features_valid)[:, 1]
    auc = roc_auc_score(target_valid, pred_prob)
    results[name] = auc
    results_df = pd.DataFrame.from_dict(results, orient = 'index', columns = ['ROC-AUC'])
    results_df.index.name = 'Model'
    results_df = results_df.sort_values('ROC-AUC', ascending = False)
results_df

ROC-AUC
Model                        
LightGBM             0.892073
XGBoost              0.890113
CatBoost             0.886889
Random Forest        0.854115
Logistic Regression  0.845545
Decision Tree        0.724422

Comparing the ROC-AUC scores of our models before cross-validation gives me a good idea of results to expect after CV. The models might not perform the exact same but this at least helps me understand which models are separating themselves from the rest.

In [18]:
models = {'Logistic Regression': LogisticRegression(class_weight = 'balanced',
                                                                    solver = 'liblinear',
                                                                    random_state = 2356
                                                   ),
          'Decision Tree': DecisionTreeClassifier(class_weight = 'balanced',
                                                  random_state = 2356
                                                 ),
          'Random Forest': RandomForestClassifier(class_weight = 'balanced',
                                                  random_state = 2356
                                                 ),
          'XGBoost': XGBClassifier(scale_pos_weight = scale_pos_weight,
                                   use_label_encoder = False,
                                   eval_metric = 'auc',
                                   random_state = 2356
                                  ),
          'LightGBM': LGBMClassifier(scale_pos_weight = scale_pos_weight,
                                     random_state = 2356
                                    ),
          'CatBoost': CatBoostClassifier(class_weights = cat_boost_weights,
                                         random_state = 2356,
                                         verbose = 0
                                        )
         }
cross_val_results = {}

for name, model in models.items():
    pipeline = Pipeline([
        ('preprocessor', preprocessor), ('classifier', model)
    ])
    scores = cross_val_score(
        pipeline,
        features_train,
        target_train,
        cv = 5,
        scoring = 'roc_auc'
    )

    cross_val_results[name] = {
        'mean_auc': scores.mean(),
        'std_auc': scores.std()
    }
cross_val_df = pd.DataFrame(cross_val_results).T
cross_val_df = cross_val_df[['mean_auc', 'std_auc']]
cross_val_df = cross_val_df.sort_values('mean_auc', ascending = False)
cross_val_df

mean_auc   std_auc
XGBoost              0.885597  0.007435
CatBoost             0.883571  0.009063
LightGBM             0.883344  0.007506
Logistic Regression  0.851195  0.014797
Random Forest        0.846320  0.013522
Decision Tree        0.696199  0.014131

After running cross-validation on the chosen models, the boosted models still performed the best, all showing mean ROC-AUC scores just over 0.88 and very minimal standard deviations. The simpler models like Logistic Regression and Random Forest performed well, and the Decision Tree Classifier was the poorest of the bunch, not even topping 0.7. For evaluating with the test set, I'll be using the 3 boosted models and comparing their performance against each other and the baseline.

## Evaluation and Analysis of Test Data

In [19]:
models = {'XGBoost': XGBClassifier(scale_pos_weight = scale_pos_weight,
                                   use_label_encoder = False,
                                   eval_metric = 'auc',
                                   random_state = 2356
                                  ),
          'LightGBM': LGBMClassifier(scale_pos_weight = scale_pos_weight,
                                     random_state = 2356
                                    ),
          'CatBoost': CatBoostClassifier(class_weights = cat_boost_weights,
                                         random_state = 2356,
                                         verbose = 0
                                        )
         }
test_results = {}

for name, model in models.items():
    pipeline = Pipeline([
        ('preprocessor', preprocessor), ('classifier', model)
    ])
    pipeline.fit(features_train, target_train)
    pred_prob = pipeline.predict_proba(features_test)[:, 1]
    auc = roc_auc_score(target_test, pred_prob)
    test_results[name] = auc
    test_results_df = pd.DataFrame.from_dict(test_results, orient = 'index', columns = ['ROC-AUC'])
    test_results_df.index.name = 'Model'
    test_results_df = test_results_df.sort_values('ROC-AUC', ascending = False)
test_results_df

ROC-AUC
Model             
CatBoost  0.867685
LightGBM  0.866738
XGBoost   0.864549

After letting the models analyze the test data, the results are very promising! The CatBoost model gets a slight edge over LGBM, with XGB not too far behind. Knowing that a score closer to 1 is indicative of a stronger model, we have 3 very high quality choices on our hands with these boosted models. 

## Report and Conclusions

Looking back at my work plan (Feature Engineering/Encoding -> Scaling and Splitting Data -> Creating Baseline and Test Models -> Model Evaluation and Selection -> Conclusion), I performed all of these steps to the best of my ability. There wasn't as much feature engineering as I initially thought, and since I'd already completed most of it in my work plan notebook, I was able to finish that step and the data scaling and splitting very quickly. As expected, most of the work and troubleshooting after came during model development and testing. Luckily we'd used these models for various other projects in the past so I was able to reference those and my notes to work through whatever issues I faced. I initially started each model separately, then after I was able to find scores for each, I was able to build the loops that tested each model at once and kept things much more organized. The Pipeline feature was new to me so I relied on some Google searching and documentation to help answer my questions about using that, and it saved me a ton of work and frustration I remember from testing models in older projects! Pipeline was definitely one of my key steps, in addition to the tips from the tutors about focusing on boosted models and how to treat their class weights after determining an imbalance earlier. 

In conclusion, I would recommend the CatBoost model as it showed the highest ROC-AUC score on the unseen test data coming in at 0.867. Slightly edging out LightGBM at 0.866, it is marginally better and if we were to add further categorical data to the set, it would be easier to interpret as we would not need to encode that data like we would for LightGBM models, saving us some time when taking the analysis even further.